File: rendezvous/internal/descriptor/descriptor.go
// Package descriptor implements the two signed descriptor types of
// RENDEZVOUS_WIRE_V1 §3.
//
// The split between PublicDescriptor and ContactDescriptor is what makes V3 §11
// ("do not expose unrestricted connection details before the recipient has
// responded") enforceable rather than aspirational: search can only ever return
// the public view, which carries no transport candidates, no IP and no port.
package descriptor

import (
	"errors"
	"strconv"

	"nex.rendezvous/internal/handle"
	"nex.rendezvous/internal/wire"
)

// Version is the only protocol version accepted in v1.
const Version = 1

// Limits from §3.2. They apply to both descriptor types where the field exists.
const (
	MaxCapabilities = 12
	MaxCapLen       = 32
	MaxCandidates   = 6
	MaxHostLen      = 255
	MaxKindLen      = 32
	MinPort         = 1
	MaxPort         = 65535
)

// MaxDescriptorLifetimeMs is the §5.1 bound: expiresAt - issuedAt <= 300000.
const MaxDescriptorLifetimeMs = 300000

// ErrInvalid marks any descriptor that fails structural validation.
var ErrInvalid = errors.New("descriptor: invalid")

// ErrBadSignature marks a descriptor whose own signature does not verify.
var ErrBadSignature = errors.New("descriptor: signature does not verify")

// ErrExpired marks a descriptor that is already past expiresAt, or whose
// lifetime exceeds the §5.1 bound.
var ErrExpired = errors.New("descriptor: expired or over-long lifetime")

// Candidate is one transport candidate. Kind is deliberately a free string:
// §3.2 requires unknown kinds to be preserved on the service side rather than
// rejected, so a later candidate kind does not need a flag day.
type Candidate struct {
	Kind string `json:"kind"`
	Host string `json:"host"`
	Port int    `json:"port"`
}

// SignInto appends this candidate's own length-prefixed fields to a signing
// input, per the §3.2 LPcand form:
//
//	LPcand(a) = uint32be(len(a)) || for each element: LP(kind) || LP(host) || LPn(port)
//
// There is no delimiter and therefore nothing an attacker-supplied kind or host
// can do to forge a field boundary.
func (c Candidate) SignInto(b *wire.Builder) {
	b.LP(c.Kind).LP(c.Host).LPn(int64(c.Port))
}

// Public is the descriptor returned by search (§3.1).
//
// No transport candidates. No IP address. No port. Search reveals that a handle
// is currently connectable and which identity claims it — nothing that lets a
// stranger dial.
type Public struct {
	V            int      `json:"v"`
	Handle       string   `json:"handle"`
	NodeID       string   `json:"nodeId"`
	SignPub      string   `json:"signPub"`
	Capabilities []string `json:"capabilities"`
	Connectable  bool     `json:"connectable"`
	IssuedAt     int64    `json:"issuedAt"`
	ExpiresAt    int64    `json:"expiresAt"`
	Sig          string   `json:"sig"`
}

// SigningInput builds the §3.1 signing input:
//
//	LP(v) LP(handle) LP(nodeId) LP(signPub) LParr(capabilities)
//	LPn(issuedAt) LPn(expiresAt) LP(connectable)
//
// v is signed as the string "1" and connectable as "true"/"false".
func (p *Public) SigningInput() []byte {
	return wire.NewBuilder(wire.DomainPublicDescriptor).
		LP(strconv.Itoa(p.V)).
		LP(p.Handle).
		LP(p.NodeID).
		LP(p.SignPub).
		LParr(p.Capabilities).
		LPn(p.IssuedAt).
		LPn(p.ExpiresAt).
		LPbool(p.Connectable).
		Bytes()
}

// Validate checks structure and then the descriptor's own signature.
//
// It does NOT prove that signPub belongs to nodeId. Nothing in this protocol
// establishes that link (§6) — the binding happens later, in the Noise
// handshake, when the dialing peer pins nodeId.
func (p *Public) Validate() error {
	if p == nil {
		return ErrInvalid
	}
	if err := validateCommon(p.V, p.Handle, p.NodeID, p.SignPub, p.Capabilities, p.Sig); err != nil {
		return err
	}
	if !wire.Verify(p.SignPub, p.SigningInput(), p.Sig) {
		return ErrBadSignature
	}
	return nil
}

// Contact is the descriptor released only after an accepted introduction (§3.2).
type Contact struct {
	V            int         `json:"v"`
	Handle       string      `json:"handle"`
	NodeID       string      `json:"nodeId"`
	SignPub      string      `json:"signPub"`
	NoisePub     string      `json:"noisePub"`
	Capabilities []string    `json:"capabilities"`
	Candidates   []Candidate `json:"candidates"`
	IssuedAt     int64       `json:"issuedAt"`
	ExpiresAt    int64       `json:"expiresAt"`
	Sig          string      `json:"sig"`
}

// SigningInput builds the §3.2 signing input:
//
//	LP(v) LP(handle) LP(nodeId) LP(signPub) LP(noisePub)
//	LParr(capabilities) LPcand(candidates) LPn(issuedAt) LPn(expiresAt)
//
// Candidates are signed in the exact array order sent.
func (c *Contact) SigningInput() []byte {
	b := wire.NewBuilder(wire.DomainContactDescriptor).
		LP(strconv.Itoa(c.V)).
		LP(c.Handle).
		LP(c.NodeID).
		LP(c.SignPub).
		LP(c.NoisePub).
		LParr(c.Capabilities).
		Count(len(c.Candidates))
	for _, cand := range c.Candidates {
		cand.SignInto(b)
	}
	return b.LPn(c.IssuedAt).LPn(c.ExpiresAt).Bytes()
}

// Validate checks structure, candidate limits, and then the signature.
func (c *Contact) Validate() error {
	if c == nil {
		return ErrInvalid
	}
	if err := validateCommon(c.V, c.Handle, c.NodeID, c.SignPub, c.Capabilities, c.Sig); err != nil {
		return err
	}
	if !wire.IsNoisePub(c.NoisePub) {
		return ErrInvalid
	}
	if len(c.Candidates) > MaxCandidates {
		return ErrInvalid
	}
	for _, cand := range c.Candidates {
		if err := validateCandidate(cand); err != nil {
			return err
		}
	}
	if !wire.Verify(c.SignPub, c.SigningInput(), c.Sig) {
		return ErrBadSignature
	}
	return nil
}

// validateCandidate applies the §3.2 limits. The kind is not checked against a
// whitelist: §3.2 requires unknown kinds to be preserved, not rejected, so a
// later candidate kind does not need a flag day.
//
// There is deliberately no delimiter-escaping rule here. Under LPcand each
// candidate field carries its own length prefix, so no kind or host value can
// forge a field boundary — the ambiguity is designed out rather than validated
// against (contract §3.2, Amendment 1).
func validateCandidate(c Candidate) error {
	if c.Kind == "" || len(c.Kind) > MaxKindLen {
		return ErrInvalid
	}
	if c.Host == "" || len(c.Host) > MaxHostLen {
		return ErrInvalid
	}
	if c.Port < MinPort || c.Port > MaxPort {
		return ErrInvalid
	}
	return nil
}

func validateCommon(v int, h, nodeID, signPub string, caps []string, sig string) error {
	if v != Version {
		return ErrInvalid
	}
	// The handle must already be in normalized form. The service re-normalizes
	// and compares rather than repairing: §2 forbids silent rewriting.
	norm, err := handle.Normalize(h)
	if err != nil || norm != h {
		return ErrInvalid
	}
	if !wire.IsNodeID(nodeID) {
		return ErrInvalid
	}
	if len(signPub) != 64 || !wire.IsLowerHex(signPub) {
		return ErrInvalid
	}
	if !wire.IsSig(sig) {
		return ErrInvalid
	}
	if len(caps) > MaxCapabilities {
		return ErrInvalid
	}
	for _, c := range caps {
		if !validCapability(c) {
			return ErrInvalid
		}
	}
	return nil
}

// validCapability enforces "at most 12 capabilities, each <= 32 bytes matching
// ^[a-z0-9-]+$" (§3.2).
func validCapability(c string) bool {
	if c == "" || len(c) > MaxCapLen {
		return false
	}
	for i := 0; i < len(c); i++ {
		b := c[i]
		if (b < 'a' || b > 'z') && (b < '0' || b > '9') && b != '-' {
			return false
		}
	}
	return true
}

// CheckLifetime applies the §5.1 freshness rules that register and refresh run
// on both descriptors: expiresAt > now, and expiresAt - issuedAt <= 300000.
func CheckLifetime(issuedAt, expiresAt, nowMs int64) error {
	if expiresAt <= nowMs {
		return ErrExpired
	}
	if expiresAt < issuedAt {
		return ErrExpired
	}
	if expiresAt-issuedAt > MaxDescriptorLifetimeMs {
		return ErrExpired
	}
	return nil
}

File: src/core/rendezvous/descriptor.ts
// Handles and signed descriptors — wire contract §2 and §3.
//
// The split between the two descriptor types is the whole reason V3 §11 ("do not
// expose connection details before the recipient has responded") is enforceable
// rather than aspirational: a search result physically cannot carry an address,
// because the object returned by search has no field for one.
import { DOMAIN, SigningInput, sign, verify, type RendezvousSigningKey } from "./framing"

/** A way to reach a node. v1 speaks only direct TCP; unknown kinds are ignored, not rejected. */
export interface TransportCandidate {
  kind: string
  host: string
  port: number
}

/** What search returns: enough to decide whether to ask, never enough to dial. */
export interface PublicDescriptor {
  v: 1
  handle: string
  nodeId: string
  signPub: string
  capabilities: string[]
  connectable: boolean
  issuedAt: number
  expiresAt: number
  sig: string
}

/** Released only after the recipient accepts an introduction. Carries addresses. */
export interface ContactDescriptor {
  v: 1
  handle: string
  nodeId: string
  signPub: string
  noisePub: string
  capabilities: string[]
  candidates: TransportCandidate[]
  issuedAt: number
  expiresAt: number
  sig: string
}

export const MAX_CANDIDATES = 6
export const MAX_CAPABILITIES = 12
const HANDLE_RE = /^[a-z0-9][a-z0-9_-]{2,31}$/
const CAPABILITY_RE = /^[a-z0-9-]{1,32}$/

/**
 * Normalize a handle, or return null when it is not a legal one.
 *
 * Deliberately returns null instead of repairing: a handle the service quietly
 * rewrote is a handle the user did not choose, and silently mapping two distinct
 * inputs onto one namespace entry is how impersonation gets in.
 */
export function normalizeHandle(raw: string): string | null {
  // ASCII whitespace only, per contract §2 step 1. JS .trim() also strips
  // Unicode whitespace (U+00A0, U+2028, ...), which the Go service does not,
  // so " roshan" normalized here but was rejected there. A handle one
  // implementation accepts and the other refuses is exactly the class of
  // silent divergence the shared conformance vectors exist to prevent.
  const trimmed = raw.replace(/^[\t\n\v\f\r ]+|[\t\n\v\f\r ]+$/g, "")
  const normalized = trimmed.normalize("NFKC").toLowerCase()
  return HANDLE_RE.test(normalized) ? normalized : null
}

function validCandidate(c: TransportCandidate): boolean {
  return (
    typeof c?.kind === "string" &&
    c.kind.length > 0 &&
    c.kind.length <= 32 &&
    typeof c.host === "string" &&
    c.host.length > 0 &&
    new TextEncoder().encode(c.host).length <= 255 &&
    Number.isInteger(c.port) &&
    c.port >= 1 &&
    c.port <= 65535
  )
}

function validCapabilities(caps: readonly string[]): boolean {
  return (
    Array.isArray(caps) && caps.length <= MAX_CAPABILITIES && caps.every((c) => CAPABILITY_RE.test(c))
  )
}

function publicInput(d: Omit<PublicDescriptor, "sig">): SigningInput {
  return new SigningInput(DOMAIN.publicDescriptor)
    .str(String(d.v))
    .str(d.handle)
    .str(d.nodeId)
    .str(d.signPub)
    .arr(d.capabilities)
    .int(d.issuedAt)
    .int(d.expiresAt)
    .bool(d.connectable)
}

function contactInput(d: Omit<ContactDescriptor, "sig">): SigningInput {
  return new SigningInput(DOMAIN.contactDescriptor)
    .str(String(d.v))
    .str(d.handle)
    .str(d.nodeId)
    .str(d.signPub)
    .str(d.noisePub)
    .arr(d.capabilities)
    .candidates(d.candidates)
    .int(d.issuedAt)
    .int(d.expiresAt)
}

export function signPublicDescriptor(
  d: Omit<PublicDescriptor, "sig">,
  key: RendezvousSigningKey,
): PublicDescriptor {
  return { ...d, sig: sign(publicInput(d), key) }
}

export function signContactDescriptor(
  d: Omit<ContactDescriptor, "sig">,
  key: RendezvousSigningKey,
): ContactDescriptor {
  return { ...d, sig: sign(contactInput(d), key) }
}

/**
 * Structural + signature validation of a descriptor received from the service.
 *
 * A valid signature here proves only that the record was not forged or mutated
 * in transit. It does NOT prove that signPub belongs to nodeId — nothing in this
 * protocol does (wire contract §6). That binding is established later, by the
 * Noise handshake, when the dialing peer pins nodeId.
 */
export function verifyPublicDescriptor(d: PublicDescriptor, now = Date.now()): boolean {
  if (!d || d.v !== 1) return false
  if (normalizeHandle(d.handle) !== d.handle) return false
  if (!/^[0-9A-F]{64}$/.test(d.nodeId)) return false
  if (typeof d.connectable !== "boolean") return false
  if (!validCapabilities(d.capabilities)) return false
  if (!Number.isInteger(d.issuedAt) || !Number.isInteger(d.expiresAt)) return false
  if (d.expiresAt <= now) return false
  return verify(publicInput(d), d.sig, d.signPub)
}

export function verifyContactDescriptor(d: ContactDescriptor, now = Date.now()): boolean {
  if (!d || d.v !== 1) return false
  if (normalizeHandle(d.handle) !== d.handle) return false
  if (!/^[0-9A-F]{64}$/.test(d.nodeId)) return false
  if (!/^[0-9a-f]{64}$/.test(d.noisePub)) return false
  if (!validCapabilities(d.capabilities)) return false
  if (!Array.isArray(d.candidates) || d.candidates.length > MAX_CANDIDATES) return false
  if (!d.candidates.every(validCandidate)) return false
  if (!Number.isInteger(d.issuedAt) || !Number.isInteger(d.expiresAt)) return false
  if (d.expiresAt <= now) return false
  return verify(contactInput(d), d.sig, d.signPub)
}

/** Dialable "host:port" strings, in candidate order. Unknown kinds are skipped. */
export function dialAddresses(d: ContactDescriptor): string[] {
  return d.candidates
    .filter((c) => c.kind === "direct-tcp")
    .map((c) => (c.host.includes(":") ? `[${c.host}]:${c.port}` : `${c.host}:${c.port}`))
}

/**
 * Punchable candidates, in published order.
 *
 * The wire contract leaves candidate kinds open and says unknown ones are
 * ignored rather than rejected, which is what lets a v1 service carry these
 * unchanged — to the service they are opaque strings inside a sealed blob it
 * cannot read anyway.
 */
export function udpCandidates(d: ContactDescriptor): Array<{ host: string; port: number }> {
  return d.candidates.filter((c) => c.kind === "udp").map((c) => ({ host: c.host, port: c.port }))
}
